agent/session.go

```diff
@@ -54,6 +56,7 @@ func newSession(ctx context.Context, agent *Agent, delay time.Duration, sessionI
 		subscriptions: make(chan *api.SubscriptionMessage),
 		registered:    make(chan struct{}),
 		closed:        make(chan struct{}),
+		cancel:        sessionCancel,
```

Comment:
Actually @dperny was correct - we do need to fork the context again.  If we don't, and we use the same cancel function, then in the case were a session times out:

The select loop in `run` (https://github.com/docker/swarmkit/pull/2134/files#diff-15ffe95a2da45f70696ffa3c01949601R89) may select `ctx.Done()` first and not write the error to `s.err`, in which case the session is not closed and rebuilt.